### ISSUE TITLE
Speeding Up Deploys via Caching

### ISSUE THREAD

**Issue by nitsujri** [customink/lamby#83]
Hi customink/@metaskills,

Thanks for this great set of gems. It really helped me wrap my head around lambda, API Gateway, SQS etc. I'm still in the learning phase so you'll have to excuse my ignorance.

At the moment, each deploy goes through a full, brand new build including installing of Gems and scratch asset precompile. 

```
# bin/_build
rm -rf ./.lamby ./.aws-sam
cp -r . "$SAM_TEMP"
mkdir -p ./.lamby
cp -r "$SAM_TEMP" "./.lamby/$RESOURCE"
pushd "./.lamby/$RESOURCE"
```

To allow caching, it can become like:

```
LAMBY_BUILD_CACHE_DIR="./.lamby/$RESOURCE"
rsync -avu --delete \
  --exclude=log \
  --exclude=node_modules \
  --exclude=vendor/bundle \
  --exclude=vendor/bundle-dev \
  --exclude=".bundle" \
  --exclude=".lamby" \
  --exclude="tmp" \
  --exclude=".git" \
  --exclude="public/packs" \
  --exclude="public/assets" \
  --exclude="config/master.key" \
  . $LAMBY_BUILD_CACHE_DIR

cd $LAMBY_BUILD_CACHE_DIR

echo '== Bundle For Deployment =='
bundle lock --add-platform x86_64-linux
bundle config --local deployment true
bundle config --local without 'development test'
bundle config --local path './vendor/bundle'
bundle install --quiet --jobs 4

echo "== Asset Hosts & Precompiling =="
NODE_ENV='production' ./bin/rails assets:precompile
```

Removing all the `rm -rf` and moving them to build specific [`dockerignore`s ](https://stackoverflow.com/questions/40904409/how-to-specify-different-dockerignore-files-for-different-builds-in-the-same-pr) allows maintaining the cache and using incremental gem-install & assets-precompile during the build stage.

First end-to-end deploy time is the same, but repeated deploys drop the time from `~15min` to `~3min` (my machine).

Obviously for larger projects with lots of gems & node packages this improves it further. Can add a `$CLEAR_CACHE` to handle busting too.

Thoughts before I dive any deeper?

**Comment by metaskills** [comment]
Generally my thoughts is that caching should be an end user concern. One that we can make easy with some guides. I'd like to do a guide on how to use GitHub Actions to do deploys for you and put an emphasis on good CI/CD practices. When doing that, I'd like to demonstrate how people might use some basic patterns with GitHub's caching to speed up deploys. Typically GitHub can deploy much faster without machine too since Docker runs so well on Linux.

So basically I'd like to write some CI/CD (with caching) guides for GitHub Actions.

**Comment by nitsujri** [comment]
I can agree with that. I'll close this thread. We can reopen if there's anything directly actionable.

**Comment by metaskills** [comment]
Gonna re-open it so I can remember to add the docs :)

**Comment by metaskills** [comment]
Got the new deploy stuff done here. https://github.com/customink/lamby-cookiecutter/pull/10 will merge shortly and get new guides written for the Lamby site.

**Comment by metaskills** [comment]
https://lamby.custominktech.com/docs/quick_start#deploy-with-github-actions
